packages/firestore/src/remote/watch_change.ts

```diff
@@ -409,16 +411,80 @@ export class WatchChangeAggregator {
         }
       } else {
         const currentSize = this.getCurrentDocumentCountForTarget(targetId);
+        // Existence filter mismatch. Remove documents to limbo, and raise a
+        // snapshot with `isFromCache:true`.
         if (currentSize !== expectedCount) {
-          // Existence filter mismatch: We reset the mapping and raise a new
-          // snapshot with `isFromCache:true`.
-          this.resetTarget(targetId);
-          this.pendingTargetResets = this.pendingTargetResets.add(targetId);
+          // Apply bloom filter to identify and move removed documents to limbo.
+          const bloomFilterApplied = this.applyBloomFilter(
+            watchChange.existenceFilter,
+            targetId,
+            currentSize
+          );
```

Comment:
when bloomFilter application is successful, mark only the removed docs `ChangeType.Removed`, they will be added to limbo directly.
Otherwise, mark all documents `ChangeType.Removed`, add this targetId to remoteEvent.targetMismatches. It will run a query, and compare the fetched documents with the marked documents to identify which ones are truly removed and put them into Limbo.